pkg/mauth/compare/compare.go

```diff
@@ -0,0 +1,334 @@
+// Package compare provides various compare operators with binary codecs
+// useful for macaroon-based authorization.
+package compare
+
+import (
+	"encoding/binary"
+	"fmt"
+	"net"
+	"regexp"
+)
+
+type Type uint8
+
+const (
+	TypeFalse    Type = 0
+	TypeTrue     Type = 1
+	TypeIntegers Type = 2
+	TypeStrings  Type = 3
+	TypeRegexp   Type = 4
+	TypeCIDRs    Type = 5
+)
+
+// comparisons are encoded as a single type byte followed by zero or more bytes
+// that are specific to the comparison type.
+//
+// UINT8 - type
+// UINT8 0..n - bytes
+
+func UnmarshalBinary(b []byte) (interface{}, error) {
+	if len(b) < 1 {
+		return nil, fmt.Errorf("compare: buffer to unmarshal is zero-length")
+	}
+	switch Type(b[0]) {
+	case TypeFalse, TypeTrue:
+		v := new(Bool)
+		return *v, v.UnmarshalBinary(b)
+	case TypeIntegers:
+		v := new(Integers)
+		return *v, v.UnmarshalBinary(b)
+	case TypeStrings:
+		v := new(Strings)
+		return *v, v.UnmarshalBinary(b)
+	case TypeRegexp:
+		v := new(Regexp)
+		return v, v.UnmarshalBinary(b)
+	case TypeCIDRs:
+		v := new(CIDRs)
+		return *v, v.UnmarshalBinary(b)
+	default:
+		return nil, fmt.Errorf("compare: unknown type %d", b[0])
+	}
+}
+
+type Bool bool
+
+func (b Bool) Compare(have bool) bool {
+	return bool(b) == have
+}
+
+func (b Bool) MarshalBinary() ([]byte, error) {
+	// type: 0x0 (false) OR 0x1 (true)
+
+	if bool(b) {
+		return []byte{byte(TypeTrue)}, nil
```

Comment:
Would it make more sense for `Type` to be `byte` rather than `uint8` to avoid `byte(TypeXXX)` everywhere?